### ISSUE TITLE
JSON formatter throws SystemStackError

### ISSUE THREAD

**Issue by raybaxter** [sidekiq/sidekiq#5858]
Ruby version: 3.0.3p157, 3.1.2p20, and 3.1.2-jemalloc
Rails version: 7.0.4.3
Sidekiq / Pro / Enterprise version(s): 7.0.7 / 7.0.9

The configuration below works under Sidekiq 7.0.6.

I believe that the error is that `Sidekiq::Logger::Formatters::JSON#call` is being passed a message that is a `Hash`. That causes an error and logging that error causes an additional error. 

Casting `message.to_s` on the following line resolves the issue in my testing.
https://github.com/sidekiq/sidekiq/blob/6192ed20b639990403f7ec03b5fd651114531368/lib/sidekiq/logger.rb#L121

The message is coming from https://github.com/sidekiq/sidekiq/blob/6192ed20b639990403f7ec03b5fd651114531368/lib/sidekiq/launcher.rb#L40

### Example `message`
```ruby
{:labels=>#<Set: {"reliable"}>, :require=>".", :environment=>nil, :concurrency=>1, :timeout=>25, :poll_interval_average=>nil, :average_scheduled_poll_interval=>5, :on_complex_arguments=>:raise, :error_handlers=>[#<Proc:0x000000012cf0b680 /Users/ray/.asdf/installs/ruby/3.0.3/lib/ruby/gems/3.0.0/gems/sidekiq-7.0.7/lib/sidekiq/config.rb:36 (lambda)>, #<Proc:0x000000011ca255b8 /Users/ray/.asdf/installs/ruby/3.0.3/lib/ruby/gems/3.0.0/gems/honeybadger-5.2.1/lib/honeybadger/plugins/sidekiq.rb:27 (lambda)>], :death_handlers=>[#<Proc:0x000000010e517778 /Users/ray/.asdf/installs/ruby/3.0.3/lib/ruby/gems/3.0.0/gems/sidekiq-pro-7.0.9/lib/sidekiq/batch/death.rb:58 (lambda)>], :lifecycle_events=>{:startup=>[], :quiet=>[#<Proc:0x000000010e515040 /Users/ray/.asdf/installs/ruby/3.0.3/lib/ruby/gems/3.0.0/gems/sidekiq-pro-7.0.9/lib/sidekiq-pro.rb:73>], :shutdown=>[#<Proc:0x000000010e72fbf0 /Users/ray/.asdf/installs/ruby/3.0.3/lib/ruby/gems/3.0.0/gems/sidekiq-pro-7.0.9/lib/sidekiq-pro.rb:45>], :heartbeat=>[#<Proc:0x0000000118bf4438 /Users/ray/code/ecfr-versioner/config/initializers/sidekiq.rb:49>, #<Proc:0x000000010e72fbc8 /Users/ray/.asdf/installs/ruby/3.0.3/lib/ruby/gems/3.0.0/gems/sidekiq-pro-7.0.9/lib/sidekiq-pro.rb:48>], :beat=>[#<Proc:0x000000012cb3c2c0 /Users/ray/.asdf/installs/ruby/3.0.3/lib/ruby/gems/3.0.0/gems/sidekiq-7.0.7/lib/sidekiq/metrics/tracking.rb:133>]}, :dead_max_jobs=>10000, :dead_timeout_in_seconds=>15552000, :reloader=>#<Sidekiq::Rails::Reloader @app=EcfrVersioner::Application>, :pidfile=>"tmp/pids/sidekiq.pid", :verbose=>true, :queues=>["versioner_1", "versioner_2", "versioner_3_indexer", "versioner_4"], :config_file=>"./config/sidekiq.yml", :backup_limit=>1000, :job_logger=>OfrJobLogger, :fetch_class=>Sidekiq::Pro::SuperFetch, :fetch_setup=>nil, :scheduled_enq=>Sidekiq::Scheduled::FastEnq, :tag=>"ecfr-versioner", :identity=>"merganser.local:85891:d96ffdf3372c"}
```

### Initializer
```ruby
SIDEKIQ_REDIS_CONNECTION_SETTINGS = {
  db: ENV["SIDEKIQ_REDIS_DB"] || 12,
  host: ENV["REDIS_PORT_6379_TCP_ADDR"] || "127.0.0.1",
  port: ENV["REDIS_PORT_6379_TCP_PORT"] || 6379
}

ENV["SIDEKIQ_REDIS_URL"] = "redis://#{SIDEKIQ_REDIS_CONNECTION_SETTINGS[:host]}:#{SIDEKIQ_REDIS_CONNECTION_SETTINGS[:port]}/#{SIDEKIQ_REDIS_CONNECTION_SETTINGS[:db]}"

Sidekiq.configure_server do |config|
  config.logger.formatter = Sidekiq::Logger::Formatters::JSON.new

  config.super_fetch!
  config.reliable_scheduler!

  config.redis = {url: ENV["SIDEKIQ_REDIS_URL"]}
end

Sidekiq::Client.reliable_push! unless Rails.env.test?
Sidekiq.configure_client do |config|
  config.redis = {url: ENV["SIDEKIQ_REDIS_URL"]}
end
Sidekiq.default_job_options = {retry: 5, backtrace: 25}

Sidekiq.strict_args!
```

### sidekiq.yml
```yaml
---
:concurrency: <%= ENV["SIDEKIQ_POOL_SIZE"] || 1 %>
:pidfile: tmp/pids/sidekiq.pid
:verbose: <%= ENV.fetch("SIDEKIQ_VERBOSITY", ENV['RAILS_ENV'] == "development" ? false : true ) %>
:queues: <%= ENV["SIDEKIQ_QUEUES"] || %w[versioner_1 versioner_2 versioner_3_indexer versioner_4] %>
```
<details>
<summary>Stack Trace</summary>

```
{"ts":"2023-03-30T01:10:11.860Z","pid":82197,"tid":"2c4x","lvl":"DEBUG","msg":"ReliablePush activated"}
{"ts":"2023-03-30T01:10:11.998Z","pid":82197,"tid":"2c4x","lvl":"INFO","msg":"Booted Rails 7.0.4.3 application in development environment"}
{"ts":"2023-03-30T01:10:11.998Z","pid":82197,"tid":"2c4x","lvl":"INFO","msg":"Running in ruby 3.0.3p157 (2021-11-24 revision 3fb7d2cadc) [arm64-darwin21]"}
{"ts":"2023-03-30T01:10:11.999Z","pid":82197,"tid":"2c4x","lvl":"INFO","msg":"Sidekiq Pro 7.0.9, commercially licensed. Thanks for your support!"}
{"ts":"2023-03-30T01:10:11.999Z","pid":82197,"tid":"2c4x","lvl":"INFO","msg":"Sidekiq 7.0.7 connecting to Redis with options {:size=>5, :pool_name=>\"internal\", :url=>\"redis://127.0.0.1:6379/12\"}"}
{"ts":"2023-03-30T01:10:12.001Z","pid":82197,"tid":"2c4x","lvl":"INFO","msg":"Sidekiq 7.0.7 connecting to Redis with options {:size=>1, :pool_name=>\"default\", :url=>\"redis://127.0.0.1:6379/12\"}"}
{"ts":"2023-03-30T01:10:12.001Z","pid":82197,"tid":"2c4x","lvl":"DEBUG","msg":"Firing startup event"}
{"ts":"2023-03-30T01:10:12.001Z","pid":82197,"tid":"2c4x","lvl":"DEBUG","msg":"Client Middleware: Sidekiq::Batch::Client"}
{"ts":"2023-03-30T01:10:12.001Z","pid":82197,"tid":"2c4x","lvl":"DEBUG","msg":"Server Middleware: Honeybadger::Plugins::Sidekiq::Middleware, Sidekiq::Metrics::Middleware, Sidekiq::Batch::Server"}
{"ts":"2023-03-30T01:10:12.001Z","pid":82197,"tid":"2c4x","lvl":"INFO","msg":"Starting processing, hit Ctrl-C to stop"}
{"ts":"2023-03-30T01:10:12.001Z","pid":82197,"tid":"2cz5","lvl":"INFO","msg":"SuperFetch[default] activated"}
bundler: failed to load command: sidekiq (/Users/ray/.asdf/installs/ruby/3.0.3/bin/sidekiq)
/Users/ray/.asdf/installs/ruby/3.0.3/lib/ruby/gems/3.0.0/gems/actionpack-7.0.4.3/lib/action_dispatch/journey/visitors.rb:162:in `visit': stack level too deep (SystemStackError)
        from /Users/ray/.asdf/installs/ruby/3.0.3/lib/ruby/gems/3.0.0/gems/actionpack-7.0.4.3/lib/action_dispatch/journey/visitors.rb:107:in `binary'
        from /Users/ray/.asdf/installs/ruby/3.0.3/lib/ruby/gems/3.0.0/gems/actionpack-7.0.4.3/lib/action_dispatch/journey/visitors.rb:109:in `visit_CAT'
        from /Users/ray/.asdf/installs/ruby/3.0.3/lib/ruby/gems/3.0.0/gems/actionpack-7.0.4.3/lib/action_dispatch/journey/visitors.rb:103:in `visit'
        from /Users/ray/.asdf/installs/ruby/3.0.3/lib/ruby/gems/3.0.0/gems/actionpack-7.0.4.3/lib/action_dispatch/journey/visitors.rb:162:in `visit'
        from /Users/ray/.asdf/installs/ruby/3.0.3/lib/ruby/gems/3.0.0/gems/actionpack-7.0.4.3/lib/action_dispatch/journey/visitors.rb:117:in `unary'
        from /Users/ray/.asdf/installs/ruby/3.0.3/lib/ruby/gems/3.0.0/gems/actionpack-7.0.4.3/lib/action_dispatch/journey/visitors.rb:119:in `visit_GROUP'
        from /Users/ray/.asdf/installs/ruby/3.0.3/lib/ruby/gems/3.0.0/gems/actionpack-7.0.4.3/lib/action_dispatch/journey/visitors.rb:103:in `visit'
        from /Users/ray/.asdf/installs/ruby/3.0.3/lib/ruby/gems/3.0.0/gems/actionpack-7.0.4.3/lib/action_dispatch/journey/visitors.rb:162:in `visit'
         ... 11352 levels...
        from /Users/ray/.asdf/installs/ruby/3.0.3/lib/ruby/gems/3.0.0/gems/bundler-2.3.26/lib/bundler/friendly_errors.rb:120:in `with_friendly_errors'
        from /Users/ray/.asdf/installs/ruby/3.0.3/lib/ruby/gems/3.0.0/gems/bundler-2.3.26/exe/bundle:36:in `<top (required)>'
        from /Users/ray/.asdf/installs/ruby/3.0.3/bin/bundle:25:in `load'
        from /Users/ray/.asdf/installs/ruby/3.0.3/bin/bundle:25:in `<main>'
```

</details>

**Comment by mperham** [comment]
Ugh yeah, that’s the code which logs the configured options on boot. I’m guessing there’s a reference to your Rails app which somehow causes it to try to parse routes.rb. I will try to reproduce it tomorrow.

**Comment by mperham** [comment]
I can't reproduce the issue. Can you give me a sample app which reproduces the error?

```
{"ts":"2023-03-30T15:12:39.645Z","pid":62989,"tid":"1e3h","lvl":"INFO","msg":"Starting processing, hit Ctrl-C to stop"}
{"ts":"2023-03-30T15:12:39.645Z","pid":62989,"tid":"1e3h","lvl":"DEBUG","msg":{"labels":"#<Set: {\"reliable\"}>","require":".","environment":null,"concurrency":5,"timeout":25,"poll_interval_average":null,"average_scheduled_poll_interval":5,"on_complex_arguments":"raise","error_handlers":["#<Proc:0x0000000105fc97d8 /Users/mperham/src/sidekiq/lib/sidekiq/config.rb:36 (lambda)>"],"death_handlers":["#<Proc:0x000000010737fa20 /Users/mperham/src/pro/lib/sidekiq/batch/death.rb:58 (lambda)>"],"lifecycle_events":{"startup":[],"quiet":["#<Proc:0x000000010737d7e8 /Users/mperham/src/pro/lib/sidekiq-pro.rb:73>"],"shutdown":["#<Proc:0x0000000107c7d820 /Users/mperham/src/pro/lib/sidekiq-pro.rb:45>","#<Proc:0x0000000107c7d320 /Users/mperham/src/pro/lib/sidekiq-pro.rb:45>"],"heartbeat":["#<Proc:0x0000000107c7d7f8 /Users/mperham/src/pro/lib/sidekiq-pro.rb:48>","#<Proc:0x0000000107c7d2f8 /Users/mperham/src/pro/lib/sidekiq-pro.rb:48>"],"beat":["#<Proc:0x0000000106375418 /Users/mperham/src/sidekiq/lib/sidekiq/metrics/tracking.rb:133>"]},"dead_max_jobs":10000,"dead_timeout_in_seconds":15552000,"reloader":"#<Sidekiq::Rails::Reloader:0x0000000107bb9380>","verbose":true,"queues":["default"],"backup_limit":1000,"fetch_class":"Sidekiq::Pro::SuperFetch","fetch_setup":null,"scheduled_enq":"Sidekiq::Scheduled::FastEnq","tag":"myapp","identity":"Mikes-MacBook-Pro.local:62989:e908ff6997e3"}}
{"ts":"2023-03-30T15:12:39.645Z","pid":62989,"tid":"182p","lvl":"INFO","msg":"SuperFetch[default] activated"}
```

**Comment by raybaxter** [comment]
Closing this as I've traced this bug to the way we are requiring `yajl-ruby` (latest 1.4.3) 

You can reproduce this in a Rails app with `:verbose: true` in `sidekiq.yml` and this require in the `Gemfile`:  

```
gem "yajl-ruby", require: "yajl/json_gem"
```

Sidekiq starts if I change that require to `"yajl"`.

**Comment by letiesperon** [comment]
@raybaxter I have a similar issue to this:

> Calling `JSON.generate(message[:reloader])` where the reloader is an instance of Sidekiq::Rails::Reloader generates the same SystemStackError.

How did you solve this?

In my case, I am using `ougai` bunyan as JSON logger formatter. Sidekiq 7 fails to launch with `SystemStackError` in the same stacktrace of the issue you posted. 
I traced it to the following line breaking, and the cause is that `data` includes a `Sidekiq::Rails::Reloader` in one of the values:

``` 
 Oj.dump(data, { mode: :custom, time_format: :xmlschema, use_as_json: true, use_to_hash: true, use_to_json: true })
```

This was working on master and I did not change `oj` version so I'm not sure how to proceed
